scripts/collectArtifacts.ts

```diff
@@ -36,201 +39,238 @@ if (args.skipDownload) {
-            Readable.fromWeb(response.body),
-            unzip.Extract({
-              path: destination,
-            }),
-          );
+
+          const downloadUrl = artifact.archive_download_url;
+          if (!downloadUrl) {
+            log.error("No download URL found");
+            return;
+          }
+          log.info`Fetching artifact URL from ${downloadUrl}`;
+
+          let innerDownloadUrl: string;
+          try {
+            innerDownloadUrl = await octokit
+              .request(
+                "GET /repos/{owner}/{repo}/actions/artifacts/{artifact_id}/{archive_format}",
+                {
+                  owner: guestRepoOwner,
+                  repo: guestRepoName,
+                  artifact_id: artifact.id,
+                  archive_format: "zip",
+                },
+              )
+              .then((response) => response.url);
+          } catch (e) {
+            if (e instanceof RequestError && e.status === 410) {
+              log.error("Artifact is expired");
+              return;
+            }
+            throw e;
+          }
+
+          const path = `${repoKey}/${
+            source.type === "branch"
+              ? `branch-${source.branch.name}`
+              : `pr-${source.pullRequest.number}`
+          }`;
+          if (args.skipDownload) {
+            log.info`Download skipped: ${innerDownloadUrl}`;
+          } else {
+            log.info`Downloading artifact from ${innerDownloadUrl}`;
+            const response = await fetch(innerDownloadUrl);
+            if (!response.ok) {
+              log.error`Failed to download artifact: ${response.statusText}`;
+              return;
+            }
+            if (!response.body) {
+              log.error("Response has no body");
+              return;
+            }
+            const destination = `${destinationDir}/${path}`;
+            log.info`Extracting artifact to ${destination}`;
+            await fs.mkdir(destination, { recursive: true });
+            await pipeline(
+              Readable.fromWeb(response.body),
+              unzip.Extract({
+                path: destination,
+              }),
+            );
+          }
+          log.info("Done.");
+
+          return { source, path };
+        } catch (e) {
+          log.error`Failed to process: ${e}`;
         }
-        log.info("Done.");
+      }),
+  );
+  const successfulDownloads = downloadTargets.filter(
+    (downloadTarget) => downloadTarget !== undefined,
+  );
+  return {
+    repoKey,
+    data: successfulDownloads,
+    numTargets: downloadTargets.length,
+  };
+};
 
-        return { source, dirname };
-      } catch (e) {
-        log.error`Failed to process: ${e}`;
-      }
-    }),
-);
-const successfulDownloads = downloadTargets.filter(
-  (downloadTarget) => downloadTarget !== undefined,
-);
-if (successfulDownloads.length === 0) {
-  throw new Error("No successful downloads");
+const successfulDownloads: Record<string, DownloadResult> = {};
```

Comment:
こうかな。
```suggestion
const successfulDownloads: Record<GuestRepoKey, DownloadResult> = {};
```
まあ`{}`の代入ができないはずなので`Partial<>`とかも要りそうな予感。